### PR TITLE
Making max integrated cache staleness configurable for Spark connector

### PR DESCRIPTION
**Pull request by jeet1995**
## Description
This PR enables the tuning of the max integrated cache staleness through the read configuration for the Spark Connector. 

## Testing
Below is some evidence of queries being cached and subsequent queries being served by the integrated cache itself (as shown by the non-zero query hit rate in orange). The idea was to issue at least a couple of queries where the subsequent query had a higher max cache staleness setting than the previous one.
image

### CODE REVIEW

File: sdk/cosmos/azure-cosmos-spark_3-1_2-12/CHANGELOG.md [hunk 1]
```diff
@@ -9,6 +9,8 @@
 #### Bugs Fixed
 
 #### Other Changes
+* Added the `spark.cosmos.read.maxIntegratedCacheStaleness.inMilliseconds` and `spark.cosmos.read.maxIntegratedCacheStaleness.inHours` configuration keys
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos-spark_3-1_2-12/CHANGELOG.md:12]: This should be in the `New Features` section in all the changelogs.

File: sdk/cosmos/azure-cosmos-spark_3_2-12/docs/configuration-reference.md [hunk 2]
```diff
@@ -38,10 +38,13 @@
...
+| `spark.cosmos.read.maxIntegratedCacheStaleness.inMilliseconds` | None    | Sets the max time window in milliseconds for which query results remain cached in the integrated cache in the dedicated gateway connectivity mode. For more information on `MaxIntegratedCacheStaleness`, please visit this [link](https://learn.microsoft.com/azure/cosmos-db/integrated-cache)                                                                                                                                                                                                                                                                                                                                                                  |
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos-spark_3_2-12/docs/configuration-reference.md:45]:
```suggestion
| `spark.cosmos.read.maxIntegratedCacheStaleness.inMilliseconds` | None    | Sets the max time window in milliseconds for which query results remain cached in the integrated cache in the dedicated gateway connectivity mode. Learn more about `MaxIntegratedCacheStaleness` [here](https://learn.microsoft.com/azure/cosmos-db/integrated-cache)                                                                                                                                                                                                                                                                                                                                                                  |
```

File: sdk/cosmos/azure-cosmos-spark_3_2-12/docs/configuration-reference.md [hunk 3]
```diff
@@ -38,10 +38,13 @@
...
+| `spark.cosmos.read.maxIntegratedCacheStaleness.inHours`        | None    | Sets the max time window in hours for which query results remain cached in the integrated cache in the dedicated gateway connectivity mode. For more information on `MaxIntegratedCacheStaleness`, please visit this [link](https://learn.microsoft.com/azure/cosmos-db/integrated-cache)                                                                                                                                                                                                                                                                                                                                                                         |
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos-spark_3_2-12/docs/configuration-reference.md:46]: Same documentation update for this as well and for other docs. 
Also, what happens if user sets both of them?

**Review by kushagraThapar — COMMENTED** [review]
Looks good overall, thanks @jeet1995 . Though I am wondering, if can we make improvement by not exposing two different configurations for the same duration?

**Comment by jeet1995** [comment]
Thanks for the review and suggestions @kushagraThapar:

> Though I am wondering, if can we make improvement by not exposing two different configurations for the same duration?
Ans: We can - an improvement I can think of is appending the unit of time to the duration when setting the config value.

> Also, what happens if user sets both of them?
Ans: At this point, if both are set, the milliseconds granularity is considered over the hours granularity.

We could discuss more offline.

File: sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/CosmosConfig.scala [hunk 4]
```diff
@@ -465,13 +474,37 @@ private object CosmosReadConfig {
...
 
+  private val MaxIntegratedCacheStalenessInMilliseconds = CosmosConfigEntry[Duration](
+    key = CosmosConfigNames.ReadMaxIntegratedCacheStalenessInMilliseconds,
+    mandatory = false,
+    defaultValue = None,
+    parseFromStringFunction = queryText => Duration.ofMillis(queryText.toInt),
+    helpMessage = cacheStalenessHelpMessageInterpolator.apply("milliseconds")
+  )
+
+  private val MaxIntegratedCacheStalenessInHours = CosmosConfigEntry[Duration](
```
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/CosmosConfig.scala:485]: please remove the inHours config

File: sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/CosmosTableSchemaInferrer.scala [hunk 5]
```diff
@@ -117,6 +118,15 @@ private object CosmosTableSchemaInferrer
       SparkUtils.safeOpenConnectionInitCaches(sourceContainer, (msg, e) => logWarning(msg, e))
       val queryOptions = new CosmosQueryRequestOptions()
       queryOptions.setMaxBufferedItemCount(cosmosInferenceConfig.inferSchemaSamplingSize)
+
+      cosmosReadConfig.maxIntegratedCacheStaleness match {
+        case CosmosConstants.invalidDuration =>
```
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/CosmosTableSchemaInferrer.scala:123]: should throw on invalid duration

File: sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/ItemsPartitionReader.scala [hunk 6]
```diff
@@ -138,6 +142,15 @@ private case class ItemsPartitionReader
         ).toInt
       )
 
+      readConfig.maxIntegratedCacheStaleness match {
```
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/ItemsPartitionReader.scala:145]: this code is duplicated in several places - please add a help method that will convert the maxINtegratedCacheStaleness into new DedicatedGatewayRequestOptions in the readConfig itself and reuse it wherever this conversion is happening

**Review by FabianMeiswinkel — CHANGES_REQUESTED** [review]
Please only expose one config option and reduce code duplication for creation of DedicatedGatewayRequestOptions

File: sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/CosmosConfig.scala [hunk 7]
```diff
@@ -465,12 +468,34 @@ private object CosmosReadConfig {
...
     val customQuery = CosmosConfigEntry.parse(cfg, CustomQuery)
     val maxItemCount = CosmosConfigEntry.parse(cfg, MaxItemCount)
     val prefetchBufferSize = CosmosConfigEntry.parse(cfg, PrefetchBufferSize)
+    val maxIntegratedCacheStalenessInMilliseconds = CosmosConfigEntry.parse(cfg, MaxIntegratedCacheStalenessInMilliseconds)
+    val dedicatedGatewayRequestOptions = {
+      val result = new DedicatedGatewayRequestOptions
+      maxIntegratedCacheStalenessInMilliseconds match {
+        case Some(stalenessProvidedByUser) =>
+          result.setMaxIntegratedCacheStaleness(stalenessProvidedByUser)
+        case None =>
```
Comment (jeet1995) [inline sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/CosmosConfig.scala:494]: @FabianMeiswinkel - I am letting the exception thrown by the parse be `CosmosConfigEntry`'s responsibility. The `case None` here acts as no-op. Let me know your thoughts.

Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/CosmosConfig.scala:494]: Yeah - looks good to me now.

**Review by FabianMeiswinkel — APPROVED** [review]
Thanks!

**Pull request merged**